libopenage/game_control.cpp

```diff
@@ -100,30 +118,102 @@ ActionMode::ActionMode(qtsdl::GuiItemLink *gui_link)
+		log::log(MSG(dbg) << "Opening military build menu");
+		Engine &engine = Engine::get();
+		engine.get_input_manager().register_context(&this->build_menu_mil_context);
+		this->announce_buttons_type();
+	});
+
+	this->build_menu_context.bind(action.get("CANCEL"), [this](const input::action_arg_t &) {
+		Engine &engine = Engine::get();
+		engine.get_input_manager().remove_context(&this->build_menu_context);
+		this->announce_buttons_type();
+	});
+
+	this->build_menu_mil_context.bind(action.get("CANCEL"), [this](const input::action_arg_t &) {
+		Engine &engine = Engine::get();
+		engine.get_input_manager().remove_context(&this->build_menu_mil_context);
+		this->announce_buttons_type();
+	});
+
 	// Villager build commands
 	// TODO place this into separate building menus instead of global hotkeys
 	auto bind_building_key = [this](input::action_t action, int building) {
-		this->bind(action, [this, building](const input::action_arg_t &) {
+		this->build_menu_context.bind(action, [this, building](const input::action_arg_t &) {
 			auto player = this->game_control->get_current_player();
 			if (this->selection->contains_builders(*player)) {
 				Engine &engine = Engine::get();
 				auto player = this->game_control->get_current_player();
 				this->type_focus = player->get_type(building);
 				if (&engine.get_input_manager().get_top_context() != &this->building_context) {
+					engine.get_input_manager().remove_context(&this->build_menu_context);
 					engine.get_input_manager().register_context(&this->building_context);
+					this->announce_buttons_type();
 				}
 			}
 		});
 	};
-	bind_building_key(action.get("BUILDING_1"), 70); // House
-	bind_building_key(action.get("BUILDING_2"), 68); // Mill
-	bind_building_key(action.get("BUILDING_3"), 584); // Mining camp
-	bind_building_key(action.get("BUILDING_4"), 562); // Lumber camp
-	bind_building_key(action.get("BUILDING_5"), 12); // barracks
-	bind_building_key(action.get("BUILDING_6"), 87); // archery range
-	bind_building_key(action.get("BUILDING_7"), 101); // stable
-	bind_building_key(action.get("BUILDING_8"), 49); // siege workshop
-	bind_building_key(action.get("BUILDING_TOWN_CENTER"), 109); // Town center
+	auto bind_building_key_mil = [this](input::action_t action, int building) {
+		this->build_menu_mil_context.bind(action, [this, building](const input::action_arg_t &) {
+			auto player = this->game_control->get_current_player();
+			if (this->selection->contains_builders(*player)) {
+				Engine &engine = Engine::get();
+				auto player = this->game_control->get_current_player();
+				this->type_focus = player->get_type(building);
+				if (&engine.get_input_manager().get_top_context() != &this->building_context) {
+					engine.get_input_manager().remove_context(&this->build_menu_mil_context);
+					engine.get_input_manager().register_context(&this->building_context);
+					this->announce_buttons_type();
+				}
+			}
+		});
+	};
```

Comment:
this seems kinda redundant with the build_menu_context binding, maybe you have an idea how we can unite both?